airbyte-integrations/connectors/source-asana/source_asana/streams.py

```diff
@@ -0,0 +1,241 @@
+#
+
+
+from abc import ABC
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional
+
+import requests
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams.http import HttpStream
+
+
+class AsanaStream(HttpStream, ABC):
+    url_base = "https://app.asana.com/api/1.0/"
+
+    primary_key = "gid"
+
+    def backoff_time(self, response: requests.Response) -> Optional[int]:
+        delay_time = response.headers.get("Retry-After")
+        if delay_time:
+            return int(delay_time)
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        decoded_response = response.json()
+        next_page = decoded_response.get("next_page")
+        if next_page:
+            return {"offset": next_page["offset"]}
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+
+        # Asana pagination could be from 1 to 100.
+        params = {"limit": 100}
+
+        params.update(self.get_opt_fields())
+
+        if next_page_token:
+            params.update(next_page_token)
+
+        return params
+
+    def get_opt_fields(self) -> MutableMapping[str, str]:
+        opt_fields = list()
+        schema = self.get_json_schema()
+
+        for prop, value in schema["properties"].items():
+            if "object" in value["type"]:
+                opt_fields.append(self._handle_object_type(prop, value))
+            elif "array" in value["type"]:
+                opt_fields.append(self._handle_array_type(prop, value.get("items", [])))
+            else:
+                opt_fields.append(prop)
+
+        return {"opt_fields": ",".join(opt_fields)} if opt_fields else dict()
+
+    def _handle_object_type(self, prop: str, value: MutableMapping[str, Any]) -> str:
+        return f"{prop}.gid"
+
+    def _handle_array_type(self, prop: str, value: MutableMapping[str, Any]) -> str:
+        if "type" in value and "object" in value["type"]:
+            return self._handle_object_type(prop, value)
+
+        return prop
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        yield from response_json.get("data", [])  # Asana puts records in a container array "data"
+
+
+class AsanaStreamWorkspacePagination(AsanaStream, ABC):
+    """
+    Few streams (Projects, Tags and Users) require passing required `workspace` argument in request.
+    So this is basically the whole point of this class - to pass `workspace` argument in request.
+    """
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        workspaces_stream = Workspaces(authenticator=self.authenticator)
```

Comment:
why can't we use sliced stream here? just override `request_params` and `stream_slice` methods

Reply:
Done.